File: M3 Intermediate DSA/src/L10_Interview_Problems_on_Arrays/Assignment3.java
package L10_Interview_Problems_on_Arrays;

public class Assignment3 {

  public int solve(int[] A) {
    int N = A.length;
    int[] prefixEven = new int[N];
    int[] prefixOdd = new int[N];
    int ans = 0;

    prefixEven[0] = A[0];

    for (int index = 1; index < N; index++) {
      if ((index & 1) == 0) {
        prefixEven[index] = prefixEven[index - 1] + A[index];
        prefixOdd[index] = prefixOdd[index - 1];
      } else {
        prefixOdd[index] = prefixOdd[index - 1] + A[index];
        prefixEven[index] = prefixEven[index - 1];
      }
    }

    if (prefixOdd[N-1] - prefixOdd[0] == prefixEven[N-1] - prefixEven[0]) {
      ans++;
    }

    for (int index = 1; index < N; index++) {
      int sumEven = prefixEven[index - 1] + prefixOdd[N-1] - prefixOdd[index];
      int sumOdd = prefixOdd[index - 1] + prefixEven[N-1] - prefixEven[index];

      if (sumEven == sumOdd) {
        ans++;
      }
    }

    return ans;
  }
}

File: M3_Intermediate_DSA/L8_Interview_Problems_on_Arrays/assignment3.go
package main

func SolveA3(A []int) int {
	N := len(A)
	ans := 0

	preEven := make([]int, N)
	preOdd := make([]int, N)

	preEven[0] = A[0]
	for i := 1; i < N; i++ {
		if (i & 1) == 0 {
			preEven[i] = preEven[i-1] + A[i]
			preOdd[i] = preOdd[i-1]
		} else {
			preOdd[i] = preOdd[i-1] + A[i]
			preEven[i] = preEven[i-1]
		}
	}

	if preEven[N-1]-preEven[0] == preOdd[N-1]-preOdd[0] {
		ans++
	}

	for i := 1; i < N; i++ {
		sumEven := preEven[i-1] + preOdd[N-1] - preOdd[i]
		sumOdd := preOdd[i-1] + preEven[N-1] - preEven[i]

		if sumEven == sumOdd {
			ans++
		}
	}

	return ans
}
